src/Microsoft.ML.Data/Prediction/Calibrator.cs

```diff
@@ -190,23 +220,27 @@ private protected void SaveCore(ModelSaveContext ctx)
             ctx.SaveModel(Calibrator, @"Calibrator");
         }
 
-        private protected static IPredictorProducing<float> GetPredictor(IHostEnvironment env, ModelLoadContext ctx)
+        private protected static TSubPredictor GetPredictor(IHostEnvironment env, ModelLoadContext ctx)
         {
-            IPredictorProducing<float> predictor;
-            ctx.LoadModel<IPredictorProducing<float>, SignatureLoadModel>(env, out predictor, ModelFileUtils.DirPredictor);
+            TSubPredictor predictor;
+            ctx.LoadModel<TSubPredictor, SignatureLoadModel>(env, out predictor, ModelFileUtils.DirPredictor);
             return predictor;
         }
 
-        private protected static ICalibrator GetCalibrator(IHostEnvironment env, ModelLoadContext ctx)
+        private protected static TCalibrator GetCalibrator(IHostEnvironment env, ModelLoadContext ctx)
         {
-            ICalibrator calibrator;
-            ctx.LoadModel<ICalibrator, SignatureLoadModel>(env, out calibrator, @"Calibrator");
+            TCalibrator calibrator;
+            ctx.LoadModel<TCalibrator, SignatureLoadModel>(env, out calibrator, @"Calibrator");
             return calibrator;
         }
     }
 
-    public abstract class ValueMapperCalibratedPredictorBase : CalibratedPredictorBase, IValueMapperDist, IFeatureContributionMapper, ICalculateFeatureContribution,
+    public abstract class ValueMapperCalibratedPredictorBase<TSubPredictor,TCalibrator> :
```

Comment:
>ValueMapperCalibratedPredictorBase [](start = 26, length = 34)

Is there any reason why this should be exposed as a public type? More broadly, is there any reason why *user* code should ever need to know that the model is anything more specific than what we currently call a `CalibratedPredictorBase`?

It just strikes me that since we have made `ValueMapper

Reply:
Let me make it internal.

---
In reply to: [254540984](https://github.com/dotnet/machinelearning/pull/2435#discussion_r254540984) [](ancestors = 254540984)